File: Ising-Exp_Criticos-master 23-3-2023/Ising.py
import numpy as np
from concurrent.futures import *
import csv
import time
from tqdm import tqdm

from Medidas import *
from Montecarlo import *
from Representacion import *


def main():

    start=time.time()   # Comienza a contar el tiempo de ejecución
    

    csv_Input = open('./Input/Input.csv', 'r') # Abrir el fichero donde están los Inputs del programa
    Reader_Input = csv.reader(csv_Input) # Definir un objeto que lee el fichero

    list_Input = [] # Vector donde guardaremos los inputs
    for row in Reader_Input:
        list_Input.append(row[1]) # Añadir cada segundo elemento de cada fila del csv en una lista
    csv_Input.close() # Cerrar el fichero

    N = eval(list_Input[0]) # Número de puntos de la red
    N_mover=eval(list_Input[1]) # Número de veces que movemos antes de medir
    N_med=eval(list_Input[2])   # Número de veces que medimos
    N_calentar=eval(list_Input[3])  # Número de movimientos de calentamiento
    B = eval(list_Input[4]) # Campo externo (B=\mu H)
    beta = eval(list_Input[5]) # Inversa de la temperatura, hemos tomado k=1
    pintar_intermedios = eval(list_Input[6])    # Variable que nos dice si pintar los pasos intermedios o no


    # Creamos un vector, llamada vec_exp, que guarda los posibles valores de e^(\beta * H)
    s0 = 1
    vec_exp = []
    for i in range (2):
        for j in range (5):
            vec_exp.append(math.exp(-2 * beta * ( s0*((j-2)*2) + B * s0)))
        s0 = s0*(-1)
    
    # Abrimos los archivos donde guardaremos las medidas
    M_file=open('./Data/Magnetizacion.csv', "w", newline="")
    M_csv=csv.writer(M_file)
    M2_file=open('./Data/Magnetizacion2.csv', "w", newline="")
    M2_csv=csv.writer(M2_file)
    e_file=open('./Data/energia_promedio.csv', "w", newline="")
    e_csv=csv.writer(e_file)
    e2_file=open('./Data/energia2_promedio.csv', "w", newline="")
    e2_csv=csv.writer(e2_file)
    E_file=open('./Data/Energia.csv', "w", newline="")
    E_csv=csv.writer(E_file)

    #   Creamos la red inicial y la pintamos
    red = red_random(N)
    pintar(red, N, "inicial")

    #   Hallamos la magnetización y magnetización al cuadrado inciales y las guardamos en los respectivos ficheros
    medida = magnetizaciones(red, N)
    M = medida[0]
    M2 = medida[1]

    #   Cerramos los archivos 
    M_csv.writerow([M])
    M2_csv.writerow([M2])

    #   Inicializamos el valor de la energía
    e_i = 0
    e2 = 0
    E_i = 0

    print("Calentamiento: ")
    #     Hacemos movimientos de calentamiento
    for j in tqdm(range(N_calentar)):
        Metropolis(red, N, B, beta, vec_exp, e_i)

    #   Caculamos la energía inicial
    E_i=E_0(red, N, B)
    E_csv.writerow([E_i])


    print("Montecarlo: ")
    #   Ejecutamos el algortimo de metropolis. Movemos N_mover entre cada medida. Medimo N_med veces
    for i in tqdm(range(N_med)):
        for j in range(N_mover):
            Lista = Metropolis(red, N, B, beta, vec_exp, E_i)
            if Lista != None:
                e_i = Lista[0]
                e2 = Lista[1]
                E_i = Lista[2]
                

        # Cada vez que midamos pintamos el estado de la red, si así lo hemos decidido antes
        if pintar_intermedios == True:
            nombre="Pasos_intermedios"
            pintar(red, N, nombre)

        #   Ejecutamos la función de magnetizaciones, que nos da un valor de la magnetización o otro de la magnetización al cuadrado.
        #   Guardamos las salidas y luego las escribimos en un archivo
        medida = magnetizaciones(red, N)    
        M = medida[0]
        M2 = medida[1]
        M_csv.writerow([M])
        M2_csv.writerow([M2])
        e_csv.writerow([e_i])
        e2_csv.writerow([e2])
        E_csv.writerow([E_i])
     
 
        # print("Progreso", (100*i)/N_med, "%", end = "\r")

    #   Cerramos los archivos en lo que hemos escrito
    M_file.close()
    M2_file.close()
    e_file.close()
    e2_file.close()
    E_file.close()

    #   Pintamos la red al final del proceso
    pintar(red, N, "final")
    #   Generamos los gráficos 
    Representacion()
    #   Llamamos a la función que calcula los promedios de las cantidades necesarias
    promedios(N, beta)

    print("------------------------------------------------------------------")
    print("Duración de la simulación: ", time.time()-start, "segundos")
    print("------------------------------------------------------------------")
    print("MAGNETO")
    print("""⠀⠀⠀⠀⠀⠀⣀⣤⣴⣶⣶⣶⣶⣦⣤⣀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⢀⣴⡿⢛⣿⣿⣿⣿⣿⣿⣿⣿⡛⢿⣦⡀⠀⠀⠀
⠀⠀⣰⣿⡟⠀⠀⣿⣿⠿⠛⠛⠿⣿⣿⠀⠀⢻⣿⣆⠀⠀
⠀⣰⣿⣿⣿⡀⠀⠹⣿⠀⠀⠀⠀⣿⠏⠀⢀⣿⣿⣿⣧⠀
⢠⣿⣿⣿⣿⣿⣦⣤⣘⣇⠀⠀⣸⣁⣤⣴⣿⣿⣿⣿⣿⡄
⢸⣿⣿⣿⣿⣿⣿⣿⣿⣿⣧⣾⣿⣿⣿⣿⣿⣿⣿⣿⣿⡇
⢸⣿⣿⣿⣿⠟⠛⠛⠻⣿⣿⣿⣿⠟⡛⠛⠻⣿⣿⣿⣿⡇
⢸⣿⣿⣿⣇⠀⠚⠛⠓⠈⠛⠛⠁⠚⠛⠓⠀⣹⣿⣿⣿⡇
⢸⣿⣿⣿⣿⣷⣤⣤⣄⠀⠀⠀⠀⣠⣤⣤⣾⣿⣿⣿⣿⡇
⢸⣿⣿⣿⣿⣿⣿⣿⣿⠀⠀⠀⠀⣿⣿⣿⣿⣿⣿⣿⣿⡇
⢸⣿⣿⣿⣿⣿⣿⣿⣿⠿⠿⠿⠿⣿⣿⣿⣿⣿⣿⣿⣿⡇
⢸⣿⣿⣿⣿⣿⣿⣿⣿⠀⠀⠀⠀⣿⣿⣿⣿⣿⣿⣿⣿⡇
⢸⣿⣿⣿⣿⣿⣿⣿⣿⣦⣤⣤⣴⣿⣿⣿⣿⣿⣿⣿⣿⡇
⠘⠻⠿⣿⣿⣿⣿⣿⣿⠛⠛⠛⠛⣿⣿⣿⣿⣿⣿⠿⠟⠃
⠀⠀⠀⠀⠀⠉⠉⠉⠁⠀⠀⠀⠀⠉⠋⠉⠉⠀⠀⠀⠀⠀""")

    

if __name__ == '__main__':
    main()

File: Ising_Julia/Ising.jl
#A mi por defecto se me ejecuta con el boton ▷ en REPL, que es lo más rápido
#La primera vez que lo ejecutas tarda un ratito en empezar a ejecutar el programa (30 segundos +-)
#El resto de veces tarda muy poco
#(Creo que tiene que abrir el terminal y mierdas)
#ACTUALIZACION
#(Probando a ejecutarlo desde el REPL de julia.exe tarda lo mismo la primera vez
#Probablemente tiene que cargar paquetes o algo, no será abrir el terminal)
#ACTUALIZACION2
#Probando tiene que hacer ambas cosas, en abrir el terminal tarda unos 8 segundos
#Y en cargar los paquetes (supongo) o mierdas unos 22 segundos

#Si lo ejecutar con Ctrl+F5 o con otra cosa no se por qué tarda en empezar a ejecutarlo
#No se, no soy experto, probablemento haya otra forma de ejecutarlo sin que tenga que cargar el terminal ni tarde tanto en empezar

#Para incluir otros archivos con funciones: include("Nombre.jl")
include("Medidas.jl")
include("Montecarlo.jl")
include("Representación.jl")

#Para usar paquetes: using NombrePaquete
#ANTES HAY QUE INSTALARLOS. PARA INSTALAR:(mirar "Notas/pkg_add.csv")
#Escribir ] en el terminal de julia de VSCode (esto abre el terminal para instalar paquetes)
#Escribir es este terminal add "NombrePaquete" (mirar "Notas/pkg_add.csv" para ver los
#paquetes que hay que instalar)
using CSV
using DataFrames
using ProgressBars
using Plots

function main()

    #Lee el archivo input, lo mete en un DataFrame y lo guarda en input. [!,2] elige todas las filas y solo la columna 2.
    #Un DataFrame es una tabla con titulos, 
    #las columnas y filas de input son lo que está "dentro" de la tabla (sin titulos)
    #Para ver que es un DataFrame se puede ejecutar println(CSV.read("Input/Input.csv", DataFrame))

    input = CSV.read("Input/Input.csv", DataFrame)[!,2]
    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #--------------------------------------------INPUT--------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------

    #Guarda cada dato en una variable y le da el tipo correspondiente
    N = Int(input[1])
    N_mover = Int(input[2]*N^2)
    N_med = Int(input[3])
    N_calentar = Int(input[4]*N^2)
    B = input[5]
    β = input[6]
    red_intermedia = Bool(input[7])

    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #------------------------------------VECTORES EXPONENCIALES-----------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------

    #Escribe en una matriz 2x5 los posibles valores de la exponencial.
    vec_exp = zeros(2,5)
    s0 = -1
    for i in 1:2
        for j in 1:5
            vec_exp[i, j] = exp(2*β*(s0*((j-3)*2)+B*s0))
        end
        s0 = -s0
    end

    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #--------------------------------------------RED INICIAL--------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------

    #Crea la red inicial, la guarda y la imprime por pantalla
    red = red_random(N)
    pintar(red, N, "inicial")

    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #--------------------------------------------CALENTAMIENTO------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------

    #Realiza N_calentar movimientos de calentamiento. ProgressBar() muestra la barra de progreso
    println("Calentamiento: ")
    #Hacemos movimientos de calentamiento
    for i in ProgressBar(1:N_calentar)
        Metropolis(red, N, vec_exp)
    end    

    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #-------------------------------------------METROPOLIS----------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------

    println("Montecarlo: ")
    #Crea 5 vectores vacíos para meter las medidas
    m, m_abs, m2, E, E2 = [], [], [], [], []
    #Ejecutamos el algortimo de metropolis. Movemos N_mover entre cada medida. Medimos N_med veces
    for i in ProgressBar(1:N_med)
        for j in 1:N_mover
            Metropolis(red, N, vec_exp)
        end

        medida_m = magnetizaciones(red, N)
    #push! hace lo mismo que append en python (añade el valor medida_m[i] al vector m)
        push!(m, medida_m[1])
        push!(m_abs, medida_m[2])
        push!(m2, medida_m[3])

        medida_E = Energia(red, N, B)
        push!(E, medida_E[1])
        push!(E2, medida_E[2])

    # Cada vez que midamos pintamos el estado de la red, si así lo hemos decidido antes
        if red_intermedia == true
            pintar(red, N, "Pasos intermedios")
        end

    end

    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #--------------------------------------------GUARDAR DATOS------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------

    #Guardamos en un DataFrame las medidas con su correspondiente título
    #Así es más fácil guardarlo y extraer datos del DataFrame
    data = DataFrame(Magnetización = m,
    Magnetización_absoluta = m_abs,
    Magnetización_cuadrada = m2,
    Energía = E,
    Energía_cuadrada = E2)
    
    #Escribimos las medidas en el archivo Medidas.csv
    CSV.write("Data/Medidas.csv",data)

    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------REPRESENTACION, RED FINAL Y PROMEDIOS-----------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------
    #---------------------------------------------------------------------------------------------------------------------

    #Representamos los datos en plots
    Representacion(data)

    #Pintamos la red al final de la simulación
    pintar(red, N, "final")

    #Calculamos los promedios de magnetizaciones y energias, sus y cap_calorifica
    promedios(N, β, data)
    
    return nothing
end


#@time function() mide el tiempo que tarda en ejecutar dicha función
#@time "Algo" function() escribe Algo y luego te dice cuanto ha tardado en ejecutarse la función
#También se puede usar:
#@time begin
#    codigo
#    codigo
#end
#Y te dice el tiempo que tarda en ejcutar el codigo entre el begin y el end

@time "Ha tardado" main()

#println() y print() son diferentes:
#println() imprime lo que hay entre paréntesis seguido de un salto de línea
#print() lo imprime sin salto de línea 
#Lo que significa que si pones
#print("Lo que sea")
#print("Lo que sea2")
#Lo imprime en la misma línea: Lo que seaLo que sea2

println()
println("------------------------------------------------------------------")
println()
println("MAGNETO")
println("""⠀⠀⠀⠀⠀⠀⣀⣤⣴⣶⣶⣶⣶⣦⣤⣀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⢀⣴⡿⢛⣿⣿⣿⣿⣿⣿⣿⣿⡛⢿⣦⡀⠀⠀⠀
⠀⠀⣰⣿⡟⠀⠀⣿⣿⠿⠛⠛⠿⣿⣿⠀⠀⢻⣿⣆⠀⠀
⠀⣰⣿⣿⣿⡀⠀⠹⣿⠀⠀⠀⠀⣿⠏⠀⢀⣿⣿⣿⣧⠀
⢠⣿⣿⣿⣿⣿⣦⣤⣘⣇⠀⠀⣸⣁⣤⣴⣿⣿⣿⣿⣿⡄
⢸⣿⣿⣿⣿⣿⣿⣿⣿⣿⣧⣾⣿⣿⣿⣿⣿⣿⣿⣿⣿⡇
⢸⣿⣿⣿⣿⠟⠛⠛⠻⣿⣿⣿⣿⠟⡛⠛⠻⣿⣿⣿⣿⡇
⢸⣿⣿⣿⣇⠀⠚⠛⠓⠈⠛⠛⠁⠚⠛⠓⠀⣹⣿⣿⣿⡇
⢸⣿⣿⣿⣿⣷⣤⣤⣄⠀⠀⠀⠀⣠⣤⣤⣾⣿⣿⣿⣿⡇
⢸⣿⣿⣿⣿⣿⣿⣿⣿⠀⠀⠀⠀⣿⣿⣿⣿⣿⣿⣿⣿⡇
⢸⣿⣿⣿⣿⣿⣿⣿⣿⠿⠿⠿⠿⣿⣿⣿⣿⣿⣿⣿⣿⡇
⢸⣿⣿⣿⣿⣿⣿⣿⣿⠀⠀⠀⠀⣿⣿⣿⣿⣿⣿⣿⣿⡇
⢸⣿⣿⣿⣿⣿⣿⣿⣿⣦⣤⣤⣴⣿⣿⣿⣿⣿⣿⣿⣿⡇
⠘⠻⠿⣿⣿⣿⣿⣿⣿⠛⠛⠛⠛⣿⣿⣿⣿⣿⣿⠿⠟⠃
⠀⠀⠀⠀⠀⠉⠉⠉⠁⠀⠀⠀⠀⠉⠋⠉⠉⠀⠀⠀⠀⠀""")
